### PR TITLE
[nonbug] Mavenize EclipseLink

### PR DESCRIPTION
**Pull request by rfelcman**
This is conversion of the EclipseLink build from Ant/Tycho build into pure Maven. It's based on master branch. Mainly it contains Maven related files (POMs, assembly descriptors) but there are some shell (bash) scripts too.
There are no any java files or other project/test resources (with some exceptions).
Where to start:
1. In the _buildsystem/mavenize_  (move there with `cd buildsystem/mavenize`) directory are preparation/cleanup scripts. Purpose is populate empty Maven modules with sources/resources from Ant/Tycho projects. Entry point is `prepare.sh` . It simply calls other `*prepare.sh` scripts. If you want completely resynchronize Maven modules content call `cleanup.sh` to remove content copied by preparation script and call `prepare.sh` script again. If you want populate/cleanup particular module like _core_ use `core_cleanup.sh` `core_prepare.sh`. There are scripts per each Maven module.
Note: `bash, rsync` utilities are required
2. In the project root is parent `pom.xml`. So try there:
`mvn clean install -DskipTests`
Why `-DskipTests` see chapters: _"DB, JEE Server property files"_ in `MAVENIZE.md` or _"6 mvn clean install - blockers/errors (Apache Derby only)"_ in `MAVENIZE.pdf`
Note: JDK 8, Maven >= 3.5 is required

For more details see: `MAVENIZE.md` and `MAVENIZE.pdf`

### CODE REVIEW

File: buildsystem/mavenize/test.properties/el-testjee.wildfly.properties [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+#JEE server client library. There are Maven artifact properties. Use Maven Central to find right values.
+jee.client.groupId=org.wildfly
+jee.client.artifactId=wildfly-ejb-client-bom
+#In case of WildFly this property is used as a substitution property in eclipselink-wildfly-module.xsl file.
+jee.client.version=15.0.1.Final
+
+#Server binaries location.
+#Just binaries are needed. cargo-maven2-plugin during maven build/tests configures server/server domain automatically in .../target/cargo directory.
+cargo.container.installation.home=/usr/local/javaApplications/wildfly-15.0.1.Final
```
Comment (lukasj) [inline buildsystem/mavenize/test.properties/el-testjee.wildfly.properties]: this should be changed to some temop dir - can be ie java.io.tmpdir, or maybe better sth like ${user.home}/.eclipselinktests

File: foundation/org.eclipse.persistence.nosql.maven/pom.xml [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            <type>test-jar</type>
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <testResources>
+            <testResource>
+                <directory>${basedir}/src/test/resources/mongo</directory>
```
Comment (lukasj) [inline foundation/org.eclipse.persistence.nosql.maven/pom.xml]: if 'mongo' directory gets renamed to 'META-INF' then this whole definition can be removed

**Comment by lukasj** [comment]
DBWS-Builder-tests require MySQL DB (and code there is really DB specific), so these should run if and only if MySQL is being used and skipped/not run on other DBs

File: sdo/eclipselink.sdo.test.server.maven/src/assembly/eclipselink.sdo.test.server.ear.xml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (lukasj) [inline sdo/eclipselink.sdo.test.server.maven/src/assembly/eclipselink.sdo.test.server.ear.xml]: this belongs under 'src/test/assembly' folder (or 'src/main/assembly')

Comment (rfelcman) [reply sdo/eclipselink.sdo.test.server.maven/src/assembly/eclipselink.sdo.test.server.ear.xml]: Yes, may be similar change will be in all JPA test modules.

File: buildsystem/mavenize/test.properties/el-testjee.wildfly.properties (hunk 1 above)
Comment (rfelcman) [reply buildsystem/mavenize/test.properties/el-testjee.wildfly.properties]: By default now is WildFly automatically downloaded by _download-maven-plugin_ (see _pom_test_server.xml_). Download related properties _installation.url_, _installation.checksum.md5_ are in parent pom in wildfly, glassfish profiles.
To enable/disable it see _el-testjee.wildfly.properties_ file (_skip.jee.server.installation_, _cargo.container.installation.home_ properties).

File: foundation/org.eclipse.persistence.nosql.maven/pom.xml (hunk 2 above)
Comment (rfelcman) [reply foundation/org.eclipse.persistence.nosql.maven/pom.xml]: Done

File: sdo/eclipselink.sdo.test.server.maven/src/assembly/eclipselink.sdo.test.server.ear.xml (hunk 3 above)
Comment (rfelcman) [reply sdo/eclipselink.sdo.test.server.maven/src/assembly/eclipselink.sdo.test.server.ear.xml]: Done. Most assembly files are in the 'src/it/assembly' directory.
it...like integration tests

**Comment by rfelcman** [comment]
> DBWS-Builder-tests require MySQL DB (and code there is really DB specific), so these should run if and only if MySQL is being used and skipped/not run on other DBs

'org.eclipse.persistence.dbws.builder' stays in main build part, but tests there 'maven-failsafe-plugin' are available only when 'mysql' profile is active.

**Comment by rfelcman** [comment]
There are new modules with test framework related classes 'org.eclipse.persistence.core.test.framework', 'org.eclipse.persistence.jpa.test.framework' -> some changes in the 'core_prepare.sh' and 'jpa_test_prepare.sh' scripts.
There are changes in dependencies too. In case of JPA server side test modules in assembly descriptors.

**Comment by rfelcman** [comment]
Most tests in the modules are now under 'src/it/...' folder. Unit tests (ASM, JPQL, MOXY, some part of CORE and some others) are still under 'src/test/...' folder.
Integration tests location is specified in the property 'integration.test.directory' in parent pom.

File: jpa/eclipselink.jpa.test.maven/pom.xml [hunk 4]
```diff
@@ -0,0 +1,2066 @@
...
+            <type>${jee.client.type}</type>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <!--Filtering enables generate/substitute test properties from Maven into *.xml files.-->
+        <testResources>
+            <testResource>
+                <directory>${integration.test.directory}/resources</directory>
```
Comment (lukasj) [inline jpa/eclipselink.jpa.test.maven/pom.xml]: would it make sense to have '${integration.test.resources}' (and '${integration.test.sources}') defined at the top level?

File: jpa/eclipselink.jpa.test.maven/pom.xml [hunk 5]
```diff
@@ -0,0 +1,2066 @@
...
+                <executions>
+                    <execution>
+                        <id>filter-test-src-TestRunnerBean</id>
+                        <phase>generate-test-sources</phase>
+                        <goals>
+                            <goal>replace</goal>
+                        </goals>
+                        <configuration>
+                            <file>${project.basedir}/${integration.test.directory}/java-templates/org/eclipse/persistence/testing/framework/server/TestRunnerBean.java</file>
+                            <outputFile>${project.build.directory}/generated-test-sources/org/eclipse/persistence/testing/framework/server/TestRunnerBean.java</outputFile>
```
Comment (lukasj) [inline jpa/eclipselink.jpa.test.maven/pom.xml]: there should be one more folder between 'generated-test-sources/org'

File: jpa/eclipselink.jpa.test.maven/pom.xml [hunk 6]
```diff
@@ -0,0 +1,2066 @@
...
+                Adds to the compiler (test scope) sources root path: .../target/generated-test-sources -->
+            <plugin>
+                <groupId>org.bsc.maven</groupId>
+                <artifactId>maven-processor-plugin</artifactId>
+                <configuration>
+                    <processors>
+                        <processor>org.eclipse.persistence.internal.jpa.modelgen.CanonicalModelProcessor</processor>
+                    </processors>
+                    <sourceDirectory>${project.basedir}/${integration.test.directory}/java/</sourceDirectory>
+                    <outputDirectory>${project.build.directory}/generated-test-sources</outputDirectory>
```
Comment (lukasj) [inline jpa/eclipselink.jpa.test.maven/pom.xml]: this should go to sth like '${project.build.directory}/generated-test-sources/modelgen' instead

File: dbws/org.eclipse.persistence.dbws.maven/pom.xml [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
+    <profiles>
+        <profile>
+            <id>java11-dependencies</id>
+            <activation>
+                <jdk>(1.11,</jdk>
```
Comment (lukasj) [inline dbws/org.eclipse.persistence.dbws.maven/pom.xml]: this should be '[11,)' instead (or '[9,)'...)

**Comment by lukasj** [comment]
the warning:
```
[WARNING] Some problems were encountered while building the effective model for org.eclipse.persistence:org.eclipse.persistence.jpa.test:jar:3.0.0-SNAPSHOT
[WARNING] 'build.plugins.plugin.(groupId:artifactId)' must be unique but found duplicate declaration of plugin org.apache.maven.plugins:maven-failsafe-plugin @ org.eclipse.persistence:org.eclipse.persistence.jpa.test:[unknown-version], /Users/lukas/development/github/eclipse-ee4j/el-maven/jpa/eclipselink.jpa.test.maven/pom.xml, line 1327, column 21
```
should be fixed

File: jpa/eclipselink.jpa.test.maven/pom.xml (hunk 4 above)
Comment (rfelcman) [reply jpa/eclipselink.jpa.test.maven/pom.xml]: OK, there are two new properties '${integration.test.sources.directory}' and '${integration.test.resources.directory}'. Property '${integration.test.directory}' still remains there. It's used in some JPA test projects.

File: jpa/eclipselink.jpa.test.maven/pom.xml (hunk 5 above)
Comment (rfelcman) [reply jpa/eclipselink.jpa.test.maven/pom.xml]: OK changed.

File: dbws/org.eclipse.persistence.dbws.maven/pom.xml (hunk 7 above)
Comment (rfelcman) [reply dbws/org.eclipse.persistence.dbws.maven/pom.xml]: OK fixed.

**Comment by lukasj** [comment]
exclude module 'org.eclipse.persistence.jpa.jaxrs.test' for now, it needs more changes which we can do later

**Comment by lukasj** [comment]
when running tests with mysql, derby should not be started/stopped - I've noticed this is happening in a module with spring tests (not sure about other but I think you know ;-) )

**Comment by lukasj** [comment]
javadoc.... running `mvn clean install -Poss-release` must pass - this is supposed to generate source bundles as well as javadoc. For now warnings and errors can  be skipped.

File: buildsystem/mavenize/test.properties/el-test.oracle.properties [hunk 8]
```diff
@@ -0,0 +1,30 @@
+# DB Connection properties
+db.driver=oracle.jdbc.OracleDriver
+db.xa.driver=oracle.jdbc.OracleDriver
+#db.xa.driver=oracle.jdbc.xa.client.OracleXADataSource
+db.url=jdbc:oracle:thin:@localhost:1521:ORCL
+#db.url=jdbc:oracle:oci:@localhost:1521:ORCL
+db.user=scott
+db.pwd=tiger
+db.platform=org.eclipse.persistence.platform.database.Oracle12Platform
+#db.platform=org.eclipse.persistence.platform.database.OraclePlatform
```
Comment (lukasj) [inline buildsystem/mavenize/test.properties/el-test.oracle.properties]: this should be default, not the 12 version from the line above

File: pom.xml [hunk 9]
```diff
@@ -0,0 +1,1246 @@
...
+                <test.properties.file>${test.mysql.properties.file}</test.properties.file>
+                <!--Used by sql-maven-plugin-->
+                <db.driver.groupId>mysql</db.driver.groupId>
+                <db.driver.artifactId>mysql-connector-java</db.driver.artifactId>
+                <db.driver.version>${mysql.version}</db.driver.version>
+                <test.skip.in-memory.db>true</test.skip.in-memory.db>
+            </properties>
+        </profile>
+        <profile>
+            <id>oracle</id>
```
Comment (lukasj) [inline pom.xml]: this profile should probably define repositories as is described at https://docs.oracle.com/middleware/1213/core/MAVEN/config_maven_repo.htm#MAVEN9014 so one can configure servers in his settings.xml and be able to build specific stuff

**Comment by lukasj** [comment]
@rfelcman asm 7.2 probably needs CQ

**Comment by mwnorman** [comment]
Congrats Radek for doing what I wanted to do ... oh 13 years ago!

**Pull request merged**